test/unit/ut_broker_security.cpp

```diff
@@ -0,0 +1,198 @@
+// Copyright Takatoshi Kondo 2021
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#include "../common/test_main.hpp"
+#include "../common/global_fixture.hpp"
+
+#include <mqtt/broker/security.hpp>
+
+BOOST_AUTO_TEST_SUITE(ut_broker_security)
+
+void load_config(MQTT_NS::broker::security &security, std::string const& value)
+{
+    std::stringstream input(value);
+    security.load_json(input);
+}
+
+BOOST_AUTO_TEST_CASE(default_config) {
+    MQTT_NS::broker::security security;
+    security.default_config();
+
+    BOOST_CHECK(security.authentication_["anonymous"].auth_method == MQTT_NS::broker::security::authentication::method::anonymous);
+    BOOST_CHECK(!security.authentication_["anonymous"].digest);
+
+    BOOST_CHECK(security.login_anonymous());
+
+    BOOST_CHECK(security.auth_pub("topic", "anonymous") == MQTT_NS::broker::security::authorization::type::allow);
+    BOOST_CHECK(security.auth_pub("sub/topic", "anonymous") == MQTT_NS::broker::security::authorization::type::allow);
+    BOOST_CHECK(security.auth_pub("sub/topic1", "anonymous") == MQTT_NS::broker::security::authorization::type::allow);
+
+    BOOST_CHECK(security.auth_sub_user(security.auth_sub("topic"), "anonymous") == MQTT_NS::broker::security::authorization::type::allow);
+    BOOST_CHECK(security.auth_sub_user(security.auth_sub("sub/topic"), "anonymous") == MQTT_NS::broker::security::authorization::type::allow);
+    BOOST_CHECK(security.auth_sub_user(security.auth_sub("sub/topic1"), "anonymous") == MQTT_NS::broker::security::authorization::type::allow);
+}
+
+BOOST_AUTO_TEST_CASE(json_load) {
+    MQTT_NS::broker::security security;
+
+    std::string value = "{\"authentication\":[{\"name\":\"u1\",\"method\":\"sha256\",\"salt\":\"salt\",\"digest\":\"694073aa885f21f4dc23af70b5d2d30dc115dcfc0c5661113ca8bab2373d741d\"},{\"name\":\"u2\",\"method\":\"client_cert\",\"field\":\"CNAME\"},{\"name\":\"u3\",\"method\":\"plain_password\",\"password\":\"mypassword\"},{\"name\":\"anonymous\",\"method\":\"anonymous\"}],\"group\":[{\"name\":\"@g1\",\"members\":[\"u1\",\"u2\",\"anonymous\"]}],\"authorization\":[{\"topic\":\"#\",\"type\":\"allow\",\"pub\":[\"@g1\"]},{\"topic\":\"#\",\"type\":\"deny\",\"sub\":[\"@g1\"]},{\"topic\":\"sub/#\",\"type\":\"allow\",\"sub\":[\"@g1\"],\"pub\":[\"@g1\"]},{\"topic\":\"sub/topic1\",\"type\":\"deny\",\"sub\":[\"u1\",\"anonymous\"],\"pub\":[\"u1\",\"anonymous\"]}]}";
```

Comment:
```suggestion
    std::string value = R"*(
{
    "authentication": [
        {
            "name": "u1",
            "method": "sha256",
            "salt": "salt",
            "digest": "694073aa885f21f4dc23af70b5d2d30dc115dcfc0c5661113ca8bab2373d741d"
        }
        ,
        {
            "name": "u2",
            "method": "client_cert",
            "field": "CNAME"
        }
        ,
        {
            "name": "u2",
            "method": "plain_password",
            "password": "mypassword"
        }
        ,
        {
            "name": "anonymous",
            "method": "anonymous"
        }
    ]
    ,
    "group": [
        {
            "name": "@g1",
            "members": ["u1", "u2", "anonymous" ]
        }
    ]
    ,
    "authorization": [
        {
            "topic": "#",
            "type": "allow",
            "pub": ["@g1"]
        }
        ,
        {
            "topic": "#",
            "type": "deny",
            "sub": ["@g1"]
        }
        ,
        {
            "topic": "sub/#",
            "type": "allow",
            "sub": ["@g1"],
            "pub": ["@g1"]
        }
        ,
        {
            "topic": "sub/topic1",
            "type": "deny",
            "sub": ["u1", "anonymous"],
            "pub": ["u1", "anonymous"]
        }
    ]
}
)*";
```

Could you use raw string literals like this?
It helps understanding.
And I added auth.json indentation rule to the Wiki. Please refer it.
https://github.com/redboltz/mqtt_cpp/wiki/Coding-Rules#writing-json

Reply:
It is possible, but a tool like jsonlint.com uses different rules:
[jsonlint](https://jsonlint.com/)

So this is inconvenient, I often use jsonlint to format the json.

Reply:
https://www.json2yaml.com/

````yaml
---
# yaml supports comments!
authentication:
- name: u1
  method: sha256
  salt: salt
  digest: 694073aa885f21f4dc23af70b5d2d30dc115dcfc0c5661113ca8bab2373d741d
- name: u2
  method: client_cert
  field: CNAME
- name: anonymous
  method: anonymous
group:
- name: "@g1"
  members:
  - u1
  - u2
  - anonymous
authorization:
- topic: "#"
  type: allow
  pub:
  - "@g1"
- topic: "#"
  type: deny
  sub:
  - "@g1"
- topic: sub/#
  type: allow
  sub:
  - "@g1"
  pub:
  - "@g1"
- topic: sub/topic1
  type: deny
  sub:
  - u1
  - anonymous
  pub:
  - u1
  - anonymous